### ISSUE TITLE
Result of IOServicePortFromCGDisplayID() has a busCount of 0

### ISSUE THREAD

**Issue by ryanrhee** [kfix/ddcctl#15]
On my system, the workaround to the deprecation of `CGDisplayIOServicePort()` breaks ddcctl. In particular:

```
io_service_t framebuffer = IOServicePortFromCGDisplayID(displayID);
IOResult result = IOFBGetI2CInterfaceCount(framebuffer, &busCount);
assert(result == KERN_SUCCESS, "Failed to get the interface count.");
```

returns `KERN_SUCCESS`, but `busCount` is 0. 

If that first line is changed to use `CGDisplayIOServicePort(displayID)` then the correct bus count is returned.

I'm on a 15" Mid-2015 MBP running 10.11.6. 

I'm not familiar with either of the functions, or with Core Graphics in general. Is there something more nuanced that needs to be done w/r/t the call to IOServicePortFromCGDisplayID()? Or is this maybe Apple's bug?

**Comment by ryanrhee** [comment]
OK, so I didn't realize that `IOServicePortFromCGDisplayID` wasn't an apple API. Oops.

The problem for me is that `CGDisplaySerialNumber(displayID)` returns the correct serial number, but IOReg doesn't have serial number information (i.e. `CFDictionaryGetValueIfPresent(info, CFSTR(kDisplaySerialNumber), (const void**)&serialNumberRef)` is `NO`).

I think we should skip serial number comparisons if IOReg doesn't have serial number information. 

I'll put up a pull request shortly.

**Comment by kfix** [comment]
Weird. Can you run `make ioreg` and post the output?
I think I have the same MacBook at $JOB. What's your model number?

**Comment by ryanrhee** [comment]
> Weird. Can you run `make ioreg` and post the output?

Assuming you meant `make displaylist` :)

https://gist.github.com/ryanrhee/345c7d106f0da2a82c16b4e774d262ac

> I think I have the same MacBook at $JOB. What's your model number?

I don't have the model number (looks like you need the original receipt to look it up?) but here's a screeshot of my system information:

![screenshot 2016-09-05 12 35 59](https://cloud.githubusercontent.com/assets/242548/18253762/74aff8e0-7365-11e6-9e7d-39b71f5e4e5a.png)

**Comment by Kalvin126** [comment]
I have tried bypassing any combination of those check to get the same result. `IOServicePortFromCGDisplayID` does not seem to return the same port as `CGDisplayIOServicePort`. That being said, this new version of ddcctl does not work for my external displays anymore. If I could get a working replacement for `CGDisplayIOServicePort`, I'd consider remaking BrightnessMenulet in Swift. 

BTW, you can find your model number/identifier by clicking system report in that menu pane.

**Comment by ryanrhee** [comment]
@Kalvin126 yup, it seems the ports are different, and the one returned from `IOServicePortFromCGDisplayID` has a bus count of 0.

> you can find your model number/identifier by clicking system report in that menu pane.

Cool, thanks. The model identifier would be `MacBookPro11,5`. I think the "model number" though is something different?

**Comment by kfix** [comment]
k, mine's an 11,3 and uses NVIDIA. I can reproduce the error.
a578a4 worked for me when I made that commit, but there's been two point releases of macOS since then.

**Comment by ryanrhee** [comment]
Maybe the deprecation alternative should be gated to the version of macOS for which it works correctly for, now that we know it's broken on the latest macOS?

**Comment by Kalvin126** [comment]
I can say the work arround produces the same results on 10.10 and 10.11

**Comment by ryanrhee** [comment]
@Kalvin126 : it's possible that it worked at 10.11.4, which is the revision I'm guessing @kfix was at when a578a4 was committed.

@kfix : do you know of any alternatives that aren't deprecated? Maybe we can expose the usage of `CGDisplayIOServicePort` but gate it behind an ifdef + compiler flag?

**Comment by ryanrhee** [comment]
ping~

**Comment by Kalvin126** [comment]
A runtime check would be better instead of ifdef.

There have been efforts to find work arounds but none have popped up. Thankfully `CGDisplayIOServicePort` is not completely removed, but we don't know when it will actually be dropped. The thing about is is that, example of this PR, is that there are many factors. Diff combinations of Mac models, OS, GPUs, and displays may or may not work. It difficult to say what to do. 

From what I can see, IOServicePortFromCGDisplayID is not reliable in returning a service port where we can use for IO2C/ddc request.

**Comment by kfix** [comment]
@Kalvin126 I agree. I'm not so sure I was testing the new routine when I thought I was. It may have never worked.
@ryanrhee Code welcome. or revert and build manually if you need ddcctl to work for now.

**Comment by ryanrhee** [comment]
@kfix I'm down to submit a pull request but wanted to know how you wanted to proceed. Should I just change the call back to `CGDisplayIOServicePort()`?

**Comment by kfix** [comment]
I have a patch coming to fix IOServicePortFromCGDisplayID() :-)

On Sep 14, 2016, at 5:55 PM, Ryan <notifications@github.com<mailto:notifications@github.com>> wrote:


@kfix<https://github.com/kfix> I'm down to submit a pull request but wanted to know how you wanted to proceed. Should I just change the call back to CGDisplayIOServicePort()?
